runtime/jcl/common/java_lang_Class.cpp

```diff
@@ -1818,6 +1824,144 @@ setNestmatesIncompatibleClassChangeError(J9VMThread *vmThread, J9Class *nestMemb
 	j9mem_free_memory(msg);
 }
 
+/**
+ * Loads, sets, verifies, & returns the claimed nest host of a class
+ *
+ * If the claimed nest host of a class is successfully loaded and verified,
+ * then the nest host field of the class is set and the nest host is returned.
+ * Otherwise, the nest host field is not set and the class itself is returned.
+ * If canThrow is true and loading/verifying the nest host fails, then an
+ * exception will be set.
+ *
+ * @param vmthread[in] the current vm thread
+ * @param clazz[in,out] class who's nest host is being acquired
+ * @param canThrow[in] true if a verification error/exception can be set & thrown
+ *
+ * @returns the nest host claimed by the class or the class itself
+ */
+static J9Class *
+loadAndCheckNestHost(J9VMThread *vmThread, J9Class *clazz, BOOLEAN canThrow) {
+	J9Class *nestHost = clazz->nestHost;
+
+	if (NULL == nestHost) {
+		J9JavaVM *vm = vmThread->javaVM;
+		J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+		PORT_ACCESS_FROM_VMC(vmThread);
+
+		J9ROMClass *romClass = clazz->romClass;
+		J9UTF8 *className = J9ROMCLASS_CLASSNAME(clazz->romClass);
+		J9UTF8 *nestHostName = J9ROMCLASS_NESTHOSTNAME(romClass);
+		BOOLEAN verified = FALSE;
+
+		char *nlsMsg = NULL;
+
+		/* If no nest host is named, class is own nest host */
+		if (NULL == nestHostName) {
+			nestHost = clazz;
+			verified = TRUE;
+		} else {
+			UDATA classLoadingFlags = 0;
+			if (TRUE == canThrow) {
+				classLoadingFlags = J9_FINDCLASS_FLAG_THROW_ON_FAIL;
+			} else {
+				classLoadingFlags = J9_FINDCLASS_FLAG_EXISTING_ONLY;
+			}
+			nestHost = vmFuncs->internalFindClassUTF8(vmThread, J9UTF8_DATA(nestHostName), J9UTF8_LENGTH(nestHostName), clazz->classLoader, classLoadingFlags);
+
+			/*
+			 * If this can throw, then we must check that the nest host is successfully
+			 * loaded by the same classloader in the same package & verifies the nest
+			 * member. Upon successful verification of the nest host class, the nest
+			 * host field is set.
+			 * If this can not throw, then verification occurs in order to determine
+			 * whether a nest host or 'this' is returned.
+			 */
+			if (NULL == nestHost) {
```

Comment:
The NLS codes (eg. `J9NLS_JCL_NESTMATES_CLASS_FAILED_TO_LOAD`) can not be easily passed around like that, due to the way that NLS error messages are looked up.

Since `j9nls_lookup_message` returns a `constant char*`, we can not declare nlsTemplate at the start of the function and set its value within an 'if' clause.

--

The current handling of NLS messages is:
- NLS .h header files are generated via `make -f buildtools.mk nls`
- The header files created contain two values for each message - a MODULE value and an ID value. They also define a macro for each message that turns `msg` into `msg_MODULE, msg_ID`
eg. looking in nls/j9jclnls.h gives
```
#define J9NLS_JCL_NEST_MEMBER_CLAIMS_DIFFERENT_NEST_HOST__MODULE 0x4a39434c
#define J9NLS_JCL_NEST_MEMBER_CLAIMS_DIFFERENT_NEST_HOST__ID 42
#define J9NLS_JCL_NEST_MEMBER_CLAIMS_DIFFERENT_NEST_HOST J9NLS_JCL_NEST_MEMBER_CLAIMS_DIFFERENT_NEST_HOST__MODULE, J9NLS_JCL_NEST_MEMBER_CLAIMS_DIFFERENT_NEST_HOST__ID
```
- When compiling the VM, the pre-processor replaces each `msg` value with a pair of values
- `j9nls_lookup_message` is a wrapper for `nls_lookup_message` (from OMR). The `j9nls_lookup_message` macro takes three arguments. The omr version takes five - the port library and four values that it receives from `j9nls_lookup_message` (since the macro is expanded prior to passing it).
- Both the module and id values must be passed to `j9nls_lookup_message` in order to succeed. This causes a compilation failure, since `j9nls_lookup_message` receives too many arguments. Passing only one causes `nls_lookup_message` to fail because it does not receive enough arguments.

This can be worked around by passing both the `msg_MODULE` and `msg_ID` values into the `nls_lookup_message` or `omrnls_lookup_message` OMR macros, but I have been told to avoid doing so - that it's preferable to use the j9 versions.

Reply:
i have updated my previous comment. added an example. you don't need to access `moduleName` and `messageNum` explicitly.

why do you need to do this?
`Since j9nls_lookup_message returns a constant char*, we can not declare nlsTemplate at the start of the function and set its value within an 'if' clause.`